contents/ai/ml.md

```diff
@@ -0,0 +1,149 @@
+
+![confusion matrix](http://scikit-learn.org/stable/_images/sphx_glr_plot_confusion_matrix_001.png)  
+(image from scikit-learn's documentation)
+
+The confusion matrix above shows that the model has predicted 6 instances of 'versicolor' to be 'virginica'. These misclassifications can be investiagated to obtain insights on both model performance and data.
+
+This process of model training and refinement continues for multiple cycles until the practitioner is content with the model's performance. Finally, the model is tested on the test set and the model's performance is recorded. The model should not be adjusted from this point onwards. This test set score is used to help select between models for production.
+
+### Production
+The best model discovered is now rewritten for production. They are usually rewritten in C++, or in lower level machine learning frameworks such as Tensorflow. Models are often retrained on a regular basis when more data is available.
+
+Techniques such as model compression must also be used to ensure that the models are within the desired size limit and perform fast enough. Deep learning models are often trained against adverserial attacks for security resasons as well. 
+
+An article detailing adverserial attacks on deep learning models can be found [here](https://blog.openai.com/adversarial-example-research/).  
+Also, here are some articles detailing the process of deploying machine learning algorithms for production.  
+Dropbox: [OCR and automatic document rotation](https://blogs.dropbox.com/tech/2017/04/creating-a-modern-ocr-pipeline-using-computer-vision-and-deep-learning/)  
+Mozilla: [DeepSpeech for voice recognition](https://hacks.mozilla.org/2017/11/a-journey-to-10-word-error-rate/)
+
+### Tips and Tricks
+Practitioners find all sorts of ways to speed up their prototyping workflow.
+One commonly used trick for Python is running `%prun fn(..)`, to profile slow running functions. Profiling helps identify inefficiencies in the code which can sometimes be quickly addressed to speed up functions.
+
+A plethora of plugins exist for Jupyter. Even Vim aficionados can get their fix of Vim on Jupyter through plugins. Links for
+[general plugins](https://github.com/ipython-contrib/jupyter_contrib_nbextensions), [Vim](https://github.com/lambdalisue/jupyter-vim-binding).
+
+### Concluding Remarks 
+The process of coding a machine learning algorithm for production is an process fairly different from regular software development. 
+
+The above guide covers only the process of bringing a machine learning algorithm from prototyping to production, with no mention of specific algorithms. This process however, is mostly similar for all machine learning algorithms, from classical to new and emerging deep learning algorithms.
```

Comment:
Can consider adding more links to other resources at the end of the resource, like resources for specific algorithms for further reading.

Reply:
I was thinking that the 'Types of ML Algorithms' section covers that already. (have some links there)
I'll try to add other sorts of links below instead.